src/librustc_mir/interpret/intern.rs

```diff
@@ -422,13 +428,16 @@ where
                 }
             }
         } else if ecx.memory.dead_alloc_map.contains_key(&alloc_id) {
-            // dangling pointer
-            throw_ub_format!("encountered dangling pointer in final constant")
+            // Codegen does not like dangling pointers, and generally `tcx` assumes that
```

Comment:
In case of constants, we could just use `std::ptr::dangling` (not in this PR, just a general idea), and make this a lint

Reply:
That makes inequal pointers equal -- not sure if that is a problem?
Also we have to be careful to preserve the original alignment of the now-dead allocation (and not use the alignment of the raw ptr type).

Reply:
the addresses of constants are not something that the user can rely on in any way. We'll need to start normalizing them anyway to fix our type system unsoundness with const generics of reference type.

Reply:
The concrete addresses, not. But whether addresses are equal, and whether they are aligned, is observable inside const-eval, so we have to preserve that.

Reply:
I don't think it is observable on stable, so we can break that.

Reply:
I don't think we should build in the assumption that pointer equality will *never* be observable on stable `const fn`. That's a *huge* commitment.

For example, I presume we do want to permit the following eventually:
```rust
unsafe const fn test_ptr_eq(ptr1: *mut i32, ptr2: *mut i32) -> bool {
  let old2 = *ptr2;
  let sentinel = (*ptr1).wrapping_add(1); // definitely inequal to what is currently in ptr1.
  *ptr2 = sentinel;
  let eq = (*ptr1 == sentinel);
  // Restore old value.
  *ptr2 = old2;
  return eq;
}
```
This doesn't work for dangling pointers, granted -- but I'm just saying, this is a complex topic, and I'd rather not commit to such a limitation prematurely.

Reply:
All I mean is that we can *break* the behaviour today. We should not permanently limit us, I agree.

Reply:
Well, let's discuss that when someone proposes a concrete change. ;)
For now, I think a hard error on dangling pointers anywhere during interning is fine.